### PR TITLE
Implement AllToAll collective op

### PR DESCRIPTION
**Pull request by shaahji**
### Description
Implement AllToAll collective op

### Motivation and Context
Follow up to/continuation of #14399

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc [hunk 1]
```diff
@@ -129,12 +129,47 @@ Status AllGather::ComputeInternal(OpKernelContext* context) const {
...
+  const auto in_shape = input_tensor->Shape();
+  const int64_t input_count = in_shape.Size();
+  auto out_shape = in_shape;
+  const int64_t element_size = input_tensor->DataType()->Size();
+  const int64_t rank_stride = input_count / group_size_;
+  const ncclDataType_t dtype = GetNcclDataType(input_tensor->DataType());
+
+  char* output_data = static_cast<char *>(context->Output(0, out_shape)->MutableDataRaw());
+
+#ifdef ORT_USE_NCCL
```
Comment (souptc) [inline onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc:149]: this file is only included when nccl enabled, so don't need this guard.

**Review by souptc — DISMISSED** [review]

File: onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc (hunk 1 above)
Comment (shaahji) [reply onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc:149]: To be consistent, I have followed the same as AllGather & AllReduce introduced last week.

**Review by pranavsharma — APPROVED** [review]

**Review by souptc — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]

**Comment by shaahji** [comment]
Opened a new PR with the branch from the host repo instead of the forked one. Pipeline runs aren't supported from remote repos preventing merges.

**Pull request closed without merging**